src/assets/Generator.Shared/ArmOperationHelpers.cs

```diff
@@ -0,0 +1,389 @@
+            ScenarioInfo info = GetScenarioInfo(originalResponse, originalUri, requestMethod, finalStateVia);
+            if ((requestMethod != RequestMethod.Put && (info.HeaderFrom == HeaderFrom.None || !info.HasLocation)) || finalStateVia == FinalStateVia.AzureAsyncOperation)
+            {
+                // Support Operation Resource: https://github.com/Azure/autorest.csharp/issues/447
+                throw clientDiagnostics.CreateRequestFailedException(originalResponse);
+            }
+
+            return new ArmOperation<T>(pipeline, clientDiagnostics, originalResponse, scopeName, info, createFinalResponse, createFinalResponseAsync);
+        }
+
+        private class ArmOperation<T> : Operation<T> where T : notnull
+        {
+            private readonly HttpPipeline _pipeline;
+            private readonly ClientDiagnostics _clientDiagnostics;
+            private readonly string _scopeName;
+            private readonly Func<Response, CancellationToken, Response<T>> _createFinalResponse;
+            private readonly Func<Response, CancellationToken, ValueTask<Response<T>>> _createFinalResponseAsync;
+            private readonly ScenarioInfo _info;
+
+            private Response _rawResponse;
+            private T _value = default!;
+            private bool _hasValue;
+            private bool _hasCompleted;
+            private bool _shouldPoll;
+
+            public ArmOperation(HttpPipeline pipeline, ClientDiagnostics clientDiagnostics, Response originalResponse, string scopeName, ScenarioInfo info,
+                Func<Response, CancellationToken, Response<T>> createFinalResponse, Func<Response, CancellationToken, ValueTask<Response<T>>> createFinalResponseAsync)
+            {
+                _rawResponse = originalResponse;
+                _pipeline = pipeline;
+                _clientDiagnostics = clientDiagnostics;
+                _scopeName = scopeName;
+                _createFinalResponse = createFinalResponse;
+                _createFinalResponseAsync = createFinalResponseAsync;
+                _info = info;
+                // When the original response has no headers, we do not start polling immediately.
+                _shouldPoll = _info.HeaderFrom != HeaderFrom.None;
+            }
+
+            public override Response GetRawResponse() => _rawResponse;
+
+            public override ValueTask<Response<T>> WaitForCompletionAsync(CancellationToken cancellationToken = default) =>
+                this.DefaultWaitForCompletionAsync(OperationHelpers.DefaultPollingInterval, cancellationToken);
+
+            public override ValueTask<Response<T>> WaitForCompletionAsync(TimeSpan pollingInterval, CancellationToken cancellationToken) =>
+                this.DefaultWaitForCompletionAsync(pollingInterval, cancellationToken);
+
+            public override async ValueTask<Response> UpdateStatusAsync(CancellationToken cancellationToken = default)
```

Comment:
Let's merge two UpdateStatus* methods into one with `bool async` parameter. Don't want to have two pieces of complicated logic like this.

Reply:
Done.